core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019-2022 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+internal class LocalDateProgressionIterator(private val iterator: IntIterator) : Iterator<LocalDate> {
+    override fun hasNext(): Boolean = iterator.hasNext()
+    override fun next(): LocalDate = LocalDate.fromEpochDays(iterator.next())
+}
+
+public open class LocalDateProgression
+internal constructor(internal val intProgression: IntProgression) : Iterable<LocalDate> {
+
+    internal constructor(
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Int
+    ) : this(IntProgression.fromClosedRange(start.toEpochDays(), endInclusive.toEpochDays(), step))
+
+    public val first: LocalDate = LocalDate.fromEpochDays(intProgression.first)
+    public val last: LocalDate = LocalDate.fromEpochDays(intProgression.last)
+
+    override fun iterator(): Iterator<LocalDate> = LocalDateProgressionIterator(intProgression.iterator())
+
+    public open fun isEmpty(): Boolean = intProgression.isEmpty()
+
+    override fun toString(): String = if (intProgression.step > 0) "$first..$last step ${intProgression.step}" else "$first downTo $last step ${intProgression.step}"
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as LocalDateProgression
+
+        return intProgression == other.intProgression
+    }
+
+    override fun hashCode(): Int {
+        return intProgression.hashCode()
+    }
+
+
+    public companion object {
+        public fun fromClosedRange(
```

Comment:
Until it's clear that this function is useful, it can be kept `internal`.

Reply:
IIRC, we don't expect this function to be useful to the public (and it looks like we're right not to: https://grep.app/search?q=.fromClosedRange%28), and instead, it's there for the compiler to generate more optimized code when people write things using the functions we do expect them to use. So, `((1..11) step 2).reversed()` becomes something like `fromClosedRange(11, 1, -2)`. See https://github.com/JetBrains/kotlin/blob/658b9e9a4b11171bd09127113e152c12aaa5597d/libraries/tools/kotlin-stdlib-gen/src/templates/Ranges.kt `LocalDate`, being outside the standard library, can't benefit from this.

@ilya-g knows more about this and could maybe correct me.

Reply:
I'll switch it to internal